test/uk/gov/hmrc/traderservices/controllers/FormFieldMappingsSpec.scala

```diff
@@ -16,34 +16,151 @@
+      epuMapping.bind(Map("" -> "1224")) should haveOnlyError[EPU]("error.epu.invalid-length")
+      epuMapping.bind(Map("" -> "12A")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "AAA")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "A12")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "1A2")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "701")) should haveOnlyError[EPU]("error.epu.invalid-number")
+      epuMapping.bind(Map("" -> "999")) should haveOnlyError[EPU]("error.epu.invalid-number")
+    }
+
+    "validate EntryNumber" in {
+      entryNumberMapping.bind(Map("" -> "000000Z")) shouldBe Right(EntryNumber("000000Z"))
+      entryNumberMapping.bind(Map("" -> "A00000Z")) shouldBe Right(EntryNumber("A00000Z"))
+      entryNumberMapping.bind(Map("" -> "")) should haveOnlyError[EntryNumber]("error.entryNumber.required")
+      entryNumberMapping.bind(Map("" -> "00000Z")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-letter-wrong-position"
+      )
+      entryNumberMapping.bind(Map("" -> "0000Z")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-letter-wrong-position"
+      )
+      entryNumberMapping.bind(Map("" -> "00000")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length"
+      )
+      entryNumberMapping.bind(Map("" -> "0")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length"
+      )
+      entryNumberMapping.bind(Map("" -> "Z")) should haveOnlyError[EntryNumber]("error.entryNumber.invalid-length")
+      entryNumberMapping.bind(Map("" -> "+")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-only-digits-and-letters"
+      )
+      entryNumberMapping.bind(Map("" -> "000000Z+")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-length"
+      )
+      entryNumberMapping.bind(Map("" -> "000+000Z")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-letter-wrong-position"
+      )
+      entryNumberMapping.bind(Map("" -> "+++++++")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-ends-with-letter",
+        "error.entryNumber.invalid-letter-wrong-position"
+      )
+      entryNumberMapping.bind(Map("" -> "++++++Z")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-letter-wrong-position"
```

Comment:
Similar to above - technically Z is in correct position, although this time it's not 'after' or 'before' any **numbers** so this may be accurate in terms of the error but again, potentially confusing for users?